### ISSUE TITLE
Attribute: make writers reek by default?

### ISSUE THREAD

**Issue by chastell** [troessner/reek#491]
Continuing the discussion from https://github.com/troessner/reek/pull/481:

The Attribute smell is disabled by default – and (in its current state) I’d say rightly so, as there are many cases where reading attributes makes a lot of sense (e.g., value objects).

@beanieboi and @phillipp rightly suggested that attribute writers (and, by extension, accessors) are more smelly than attribute readers, and we might want to consider making them reek by default.

Thus, two questions:
* should attribute writers reek by default?
* if so, should it be a separate smell or a configuration on the existing Attribute smell?

My tentative answers are ‘yes’ and ‘a configuration on the existing Attribute smell’, but they’re _very_ tentative. :)

**Comment by phillipp** [comment]
I think we got to the conclusion that a seperate smell would be easier to seperate from the original smell  (like in tools and configs) and is a valid edge case of the original small.

**Comment by troessner** [comment]
>>
Thus, two questions:
should attribute writers reek by default?
if so, should it be a separate smell or a configuration on the existing Attribute smell?
My tentative answers are ‘yes’ and ‘a configuration on the existing Attribute smell’, but they’re very tentative. :)

I'm with @chastell on this one.:)

**Comment by chastell** [comment]
> we got to the conclusion that a seperate smell would be easier to seperate from the original smell (like in tools and configs) and is a valid edge case of the original small

Ah, right, and I mentioned the UtilityFunction vs FeatureEnvy case. :)

I thought about this a bit more and think that the ‘this has nothing to do with this class whatsoever (and can be moved _anywhere_ just by cutting and pasting)’ vs ‘this has less to do with the current class than this other object (so maybe should be moved _there_, but will require edits)’ difference between UtilityFunction and FeatureEnvy is much bigger than the reader vs writer distinction of attribute access. If I were looking for a smell that handles attribute readers or writers I’d look for a common one.

I wonder how many codebases have _quite valid_ attribute writers… I’d vote for ‘let’s make it configurable, enable flagging attribute writers by default and find out’. :) @phillipp / @beanieboi: wanna give it a try?

**Comment by beanieboi** [comment]
looks good to me, i would suggest something like the following for the config:

```
Attribute:
  enabled: false
  exclude: []
  ignore: :getter | :setter
```

what do you think?

**Comment by chastell** [comment]
I think I prefer ‘reader’ and ‘writer’ (as these are the words used in `attr_*` methods).

UncommunicativeVariableName uses ‘accept’ as a key for acceptable smells (by default `_` as acceptable), so maybe something like this by default?

```YAML
Attribute:
  enabled: true
  exclude: []
  accept:
  - :reader
```

**Comment by mvz** [comment]
I would want the settings to reflect that asking Reek to only flag readers is not an option. So, in particular, we don't want to allow this:

```yaml
Attribute:
  accept:
  - :writer
```

**Comment by chastell** [comment]
Ah, good point. So `accept` either `:reader` or `:accessor`? (Although the latter is the same as `enabled: false`…)

**Comment by troessner** [comment]
I retract what I said before. 
I think this is going into the wrong direction. We should make smells easier, not more complex.
I believe we should rather change the existing smell to:

-  warn on attr_writer and attr_accessor
- not on attr_reader
- the only configuration would still be on or off

WDYT?

**Comment by beanieboi** [comment]
+1

that was our initial idea.

**Comment by mvz** [comment]
@troessner :+1: It really makes no sense to ever report readers as smelly.

**Comment by troessner** [comment]
>> that was our initial idea.

Ah, ok, I kind of lost track here.:D

**Comment by beanieboi** [comment]
i think we didn’t wrote it down, but we discussed it with Piotr.
his argument was that, attr_reader could still be a smell because it embraces that you ask the object a lot in favor of tell don’t ask.
i can see both sides, i still think that declaring readers as smell is a bit to too drastic.

**Comment by chastell** [comment]
:+1: – I’m ok with not reporting readers as smells, and I think writers and accessors should be smells.

If I ever feel any stronger about reporting readers I’ll come up with a PR and a config option and we can come back to this discussion. :smile: 

@beanieboi / @phillipp: let us know whether you’d like to work on this!

**Comment by beanieboi** [comment]
yay! at this moment i'm just not sure if we should refactor the attributes spec before we change the smell itself. i'm in favour of cleaning up the spec first, it's not really nice to work with.

WDYT?

**Comment by troessner** [comment]
Yep, spec cleanup first sounds reasonable.;)

**Comment by beanieboi** [comment]
ok, the spec is refactored, now we can move forward here.

tl;dr; 
- don't report readers as smell
- enable attribute smell by default?

anybody wants to veto? if not i'll prep a PR for this. :dancers:

**Comment by troessner** [comment]
>>
don't report readers as smell
enable attribute smell by default?

Absolutely yes, please go for it.:)

**Comment by troessner** [comment]
Doesn't seem like @beanieboi is still on that. Anybody wants to take over?

**Comment by mvz** [comment]
I'm up for taking this over.

**Comment by troessner** [comment]
:1

**Comment by troessner** [comment]
#613 is merged, so I'm closing this one.
